walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -15,9 +15,124 @@
 package com.google.sps;
 
 import java.util.Collection;
+import java.util.Collections;
+import java.util.ArrayList;
+import java.util.List;
 
 public final class FindMeetingQuery {
-  public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
-  }
+
+	private static final int END_OF_DAY = 1440;
+
+	public Collection<TimeRange> query(Collection<Event> events,
+		MeetingRequest request) {
+		int duration = (int) request.getDuration();
+		Collection<String> attendees = request.getAttendees();
+		Collection<String> optionalAttendees = request.getOptionalAttendees();
+
+		Collection<TimeRange> availableTimes = new ArrayList<>();
+		Collection<TimeRange> availableTimesWithOptionals = new ArrayList<>();
+
+		availableTimes = getAvaliableTimes(attendees, events, duration);
+
+		if (!optionalAttendees.isEmpty()) {
+			availableTimesWithOptionals = getAvaliableTimes(optionalAttendees,
+				events, duration);
+		}
+
+		if (optionalAttendees.isEmpty()) {
+			return availableTimes;
+		}
+
+		if (attendees.isEmpty()) {
+			return availableTimesWithOptionals;
+		}
+
+		availableTimesWithOptionals = intersection(availableTimesWithOptionals,
+			availableTimes);
+
+		if (!availableTimesWithOptionals.isEmpty()) {
+			return availableTimesWithOptionals;
+		}
+
+		return availableTimes;
```

Comment:
Add a comment about why you only return availableTimesWithOptionals if it's not empty.